### PR TITLE
Java11 upgrade

### PR DESCRIPTION
**Pull request by gregorgabriel**
- upgrade to java 11
- make use of java.net.http.HttpClient instead of apache's library

### CODE REVIEW

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 1]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
...
-                System.out.println("Error..? " + response.getStatusLine().getReasonPhrase());
-                throw new IOException(response.getStatusLine().getReasonPhrase());
-            }
+        HttpRequest request = HttpRequest.newBuilder()
+                .GET()
+                .version(HttpClient.Version.HTTP_2)
+                .uri(getUrl)
+                .timeout(TIMEOUT_DURATION)
+                .build();
+        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/BareConnection.java]: Something like:

```
httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(HttpResponse::body)
                .thenApply(body -> GSON.fromJson(body), BareAuthority.class)
                .exceptionally(IOException::new)
                .get();
```

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 2]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
...
+        final String body = new Gson().toJson(authorityIdentifier, AuthorityIdentifier.class);
+        HttpRequest.BodyPublisher bodyPublisher = HttpRequest.BodyPublishers.ofString(body);
         String apiKeyAuth = APIKEY_KEY + SPACE + Config.getInstance().getBareApikey();
-        httpPost.addHeader("Authorization", apiKeyAuth);
-        httpPost.addHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
-        httpPost.setEntity(new StringEntity(new Gson().toJson(authorityIdentifier, AuthorityIdentifier.class)));
-        System.out.println("httpPost=" + httpPost);
-        return httpClient.execute(httpPost);
+
+        HttpRequest request = HttpRequest.newBuilder()
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/BareConnection.java]: Possibly extract as method for reuse below

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 3]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
...
         URI uri = new URIBuilder()
                 .setScheme(HTTPS)
                 .setHost(Config.getInstance().getBareHost())
                 .setPathSegments(PATH_SEGMENT_AUTHORITY, PATH_SEGMENT_REST, PATH_SEGMENT_AUTHORITIES, PATH_SEGMENT_V_2,
                         authoritySystemControlNumber, PATH_SEGMENT_IDENTIFIERS)
                 .build();
         System.out.println("uri=" + uri);
-        HttpPost httpPost = new HttpPost(uri);
 
+        final String body = new Gson().toJson(authorityIdentifier, AuthorityIdentifier.class);
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/BareConnection.java]: Reuse a ```private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();```

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java [hunk 4]
```diff
@@ -67,34 +65,33 @@ protected GatewayResponse createAuthorityOnBare(String name) {
...
-                        gatewayResponse.setBody(gson.toJson(authorities, authorityListType));
-                        gatewayResponse.setStatusCode(Response.Status.OK.getStatusCode());
-                    } else {
-                        System.out.println(String.format(COMMUNICATION_ERROR_WHILE_CREATING, name));
-                        gatewayResponse.setErrorBody(String.format(COMMUNICATION_ERROR_WHILE_CREATING, name));
-                        gatewayResponse.setStatusCode(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
-                    }
+            if (response.statusCode() == Response.Status.CREATED.getStatusCode()
+                    || response.statusCode() == Response.Status.OK.getStatusCode()) { //201
+                BareAuthority createdAuthority = new Gson().fromJson(response.body(), BareAuthority.class);
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: Can this be handled by CompletableFuture::thenApply, see above

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java [hunk 5]
```diff
@@ -67,34 +65,33 @@ protected GatewayResponse createAuthorityOnBare(String name) {
...
-                        Type authorityListType = new TypeToken<ArrayList<Authority>>(){}.getType();
-                        Gson gson = new Gson().newBuilder().setPrettyPrinting().create();
-                        gatewayResponse.setBody(gson.toJson(authorities, authorityListType));
-                        gatewayResponse.setStatusCode(Response.Status.OK.getStatusCode());
-                    } else {
-                        System.out.println(String.format(COMMUNICATION_ERROR_WHILE_CREATING, name));
-                        gatewayResponse.setErrorBody(String.format(COMMUNICATION_ERROR_WHILE_CREATING, name));
-                        gatewayResponse.setStatusCode(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
-                    }
+            if (response.statusCode() == Response.Status.CREATED.getStatusCode()
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: The status management can also be handled in the fluent API

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java [hunk 6]
```diff
@@ -67,34 +65,33 @@ protected GatewayResponse createAuthorityOnBare(String name) {
...
+                    || response.statusCode() == Response.Status.OK.getStatusCode()) { //201
+                BareAuthority createdAuthority = new Gson().fromJson(response.body(), BareAuthority.class);
+                if (Objects.nonNull(createdAuthority)) {
+                    final Authority authority = authorityConverter.asAuthority(createdAuthority);
+                    List<Authority> authorities = new ArrayList<>();
+                    authorities.add(authority);
+                    Type authorityListType = new TypeToken<ArrayList<Authority>>() {}.getType();
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: Can we dispose with the type by using Authority[]?

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 7]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
...
-                    BareAuthority fetchedAuthority = new Gson().fromJson(streamReader, BareAuthority.class);
-                    return fetchedAuthority;
-                }
-            } else {
-                System.out.println("Error..? " + response.getStatusLine().getReasonPhrase());
-                throw new IOException(response.getStatusLine().getReasonPhrase());
-            }
+        HttpRequest request = HttpRequest.newBuilder()
+                .GET()
+                .version(HttpClient.Version.HTTP_2)
```
[the remark this answers is not in the document]
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: Default is 2.0 and fallback is version 1.1 anyway. So I remove all explicit version(HTTP_2).

[httpClient.verion()](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html#version())

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 8]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
         URI uri = new URIBuilder()
```
[the remark this answers is not in the document]
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: I guess I drop the whole method and move to the only place it is used.

File: src/main/java/no/unit/nva/bare/BareConnection.java [hunk 9]
```diff
@@ -78,63 +73,75 @@ protected URI generateGetUrl(String systemControlNumber)
...
                 .setScheme(HTTPS)
                 .setHost(Config.getInstance().getBareHost())
-                .setPath(Config.BARE_GET_PATH + "/" + systemControlNumber)
+                .setPath(Config.BARE_GET_PATH + PATH_SEPARATOR + systemControlNumber)
```
[the remark this answers is not in the document]
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: Yes, lets do that.

File: src/main/java/no/unit/nva/bare/BareConnection.java (hunk 3 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: yes

File: src/main/java/no/unit/nva/bare/BareConnection.java (hunk 2 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: Acctually I do not see how we could gain here?!
Maybe if we take only the requestBuilder in a extra method. But we sitt have to do the POST(), GET(), etc. pp. in the method itself:

`final HttpRequest.Builder requestBuilder = getHttpRequestBuilder(uri);
        HttpRequest request = requestBuilder.POST(bodyPublisher).build();`

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java (hunk 6 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: Ok lets try

File: src/main/java/no/unit/nva/bare/BareConnection.java (hunk 1 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/BareConnection.java]: That is a valid suggestion. It leads to a different responsibilties in the classes.
I think taht is out of scope of that java 11 upgrading pull request.

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java (hunk 4 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: Good idea but it leads to a different responsibilties in the classes.
I think taht is out of scope of that java 11 upgrading pull request.

File: src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java (hunk 5 above)
Comment (gregorgabriel) [reply src/main/java/no/unit/nva/bare/CreateAuthorityHandler.java]: I do not really get this comment but it is out of scope of the java 11 upgrade pull request.

**Pull request merged**